include/rocksdb/trace_record.h

```diff
@@ -0,0 +1,188 @@
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#include <string>
+#include <unordered_map>
+#include <vector>
+
+#include "rocksdb/rocksdb_namespace.h"
+#include "rocksdb/status.h"
+
+// To do: ROCKSDB_LITE ?
+
+namespace ROCKSDB_NAMESPACE {
+
+class ColumnFamilyHandle;
+class DB;
+
+// Supported trace record types.
+enum TraceType : char {
+  kTraceNone = 0,
+  kTraceBegin = 1,
+  kTraceEnd = 2,
+  // Query level tracing related trace types.
+  kTraceWrite = 3,
+  kTraceGet = 4,
+  kTraceIteratorSeek = 5,
+  kTraceIteratorSeekForPrev = 6,
+  // Block cache tracing related trace types.
+  kBlockTraceIndexBlock = 7,
+  kBlockTraceFilterBlock = 8,
+  kBlockTraceDataBlock = 9,
+  kBlockTraceUncompressionDictBlock = 10,
+  kBlockTraceRangeDeletionBlock = 11,
+  // IO tracing related trace type.
+  kIOTracer = 12,
+  // Query level tracing related trace type.
+  kTraceMultiGet = 13,
+  // All trace types should be added before kTraceMax
+  kTraceMax,
+};
+
+// Base class for all types of trace records.
+class TraceRecord {
+ public:
+  TraceRecord();
+  explicit TraceRecord(uint64_t ts);
+  virtual ~TraceRecord();
+
+  virtual TraceType GetType() const = 0;
+
+  // Execute the trace record on the specified DB and handles.
+  virtual Status Execute(DB* /*db*/,
+                         const std::vector<ColumnFamilyHandle*>& /*handles*/) {
+    return Status::NotSupported();
+  }
+  virtual Status Execute(
+      DB* /*db*/,
+      const std::unordered_map<uint32_t, ColumnFamilyHandle*>& /*cf_map*/) {
+    return Status::NotSupported();
+  }
+
+  // Timestamp (in microseconds) of this trace.
+  uint64_t timestamp;
+};
+
+// Base class for all query types of trace records.
+class QueryTraceRecord : public TraceRecord {
+ public:
+  QueryTraceRecord();
+  explicit QueryTraceRecord(uint64_t ts);
+  virtual ~QueryTraceRecord() override;
+};
+
+// Trace record for DB::Write() operation.
+class WriteQueryTraceRecord : public QueryTraceRecord {
+ public:
+  WriteQueryTraceRecord();
+  WriteQueryTraceRecord(const std::string& rep, uint64_t ts);
+  WriteQueryTraceRecord(std::string&& rep, uint64_t ts);
+  virtual ~WriteQueryTraceRecord() override;
+
+  TraceType GetType() const override { return kTraceWrite; };
+
+  virtual Status Execute(
+      DB* db, const std::vector<ColumnFamilyHandle*>& handles) override;
+  virtual Status Execute(
+      DB* db,
+      const std::unordered_map<uint32_t, ColumnFamilyHandle*>& cf_map) override;
+
+  // Serialized string object to construct a WriteBatch.
+  std::string rep;
+};
+
+// Trace record for DB::Get() operation
+class GetQueryTraceRecord : public QueryTraceRecord {
+ public:
+  GetQueryTraceRecord();
+  GetQueryTraceRecord(uint32_t cf_id, const std::string& key, uint64_t ts);
+  GetQueryTraceRecord(uint32_t cf_id, std::string&& key, uint64_t ts);
```

Comment:
This pattern is fine to save a string move construction. But wouldn't it be more beneficial to avoid per-field buffers entirely? We should at least try to avoid exposing per-field buffers in the public interface because then we're basically stuck supporting them.

I can see it's non-trivial because buffers are copied from `Trace::payload` which goes out of scope before `ReplayerImpl::Next()` returns. Could `TraceRecord`s take ownership of the underlying storage buffer (`Trace::payload`) and then only expose `Slice`s in its public interface?

Reply:
That's the good point. I had several discussion with Merlin in terms of using Slice or using string. If we consider to provide full flexibility to the user, then user should be able to exam and make the change to the trace record content before apply it to Execute. So, using string which carry the ownership may be a better solution.

But I thought it again today. Actually, if the user want to change any content of the trace record and apply the updated trace record in `Execute`, user can create its own trace record, which reference to the content that user hold the ownership (user create the key or value string and reference them in the new trace record). Therefore, in this case, Slice still works. 

@autopear I agree with Andrew that we should use Slice in the public interface. The ownership of the original trace record should be the storage layer. It is also more efficient.

Reply:
Done